src/DracoCoffeeManager.php

```diff
@@ -0,0 +1,188 @@
+ */
+class DracoCoffeeManager {
+
+  /**
+   * The config factory.
+   *
+   * @var \Drupal\Core\Config\ConfigFactoryInterface
+   */
+  protected $configFactory;
+
+  /**
+   * The Entity Type Manager service.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The State API service.
+   *
+   * @var \Drupal\Core\State\StateInterface
+   */
+  protected $state;
+
+  /**
+   * The time service.
+   *
+   * @var \Drupal\Component\Datetime\TimeInterface
+   */
+  protected $time;
+
+  /**
+   * Constructs a DracoCoffeeManager object.
+   *
+   * @param \Drupal\draco_coffee\ConfigFactoryInterface $config_factory
+   *   The config factory.
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface
+   *   The Entity Type Manager service.
+   * @param \Drupal\Core\State\StateInterface $state
+   *   The State API service.
+   * @param \Drupal\Component\Datetime\TimeInterface $time
+   *   The Time service.
+   */
+  public function __construct(ConfigFactoryInterface $config_factory, EntityTypeManagerInterface $entity_type_manager, StateInterface $state, TimeInterface $time) {
+    $this->configFactory = $config_factory;
+    $this->entityTypeManager = $entity_type_manager;
+    $this->state = $state;
+    $this->time = $time;
+  }
+
+  /**
+   * Cron implementation.
+   */
+  public function cron() {
+    $config = $this->configFactory->get('draco_coffee.settings');
+    if (empty($this->state->get('draco_coffee.start')) ||
+      empty($config->get('role')) ||
+      empty($config->get('pots')) ||
+      empty($config->get('interval'))) {
+      return;
+    }
+
+    if ($this->isRefillNeeded()) {
+      $this->setBarista();
+      $this->increasePotCounter();
+
+    }
+    else {
+      $this->clear();
+    }
+    \Drupal::service('cache_tags.invalidator')->invalidateTags(['draco_coffee:state']);
+  }
+
+  /**
+   * Checks if the pot needs to be refilled.
+   *
+   * We compare the number of pots already made against the configured value,
+   * and then we check if there are remaining pots to make by comparing
+   * the time of the next pot with the current time.
+   *
+   * @return bool
+   *   TRUE if a new pot is needed. FALSE otherwise.
+   */
+  protected function isRefillNeeded() {
+    $config = $this->configFactory->get('draco_coffee.settings');
+    $next_pot = $this->state->get('draco_coffee.start') + ($this->state->get('draco_coffee.pot_counter') * 3600);
+    return (($this->state->get('draco_coffee.pot_counter') < $config->get('pots')) &&
+      ($next_pot < $this->time->getCurrentTime()));
+  }
+
+  /**
+   * Sets the editor to become barista for the next coffee pot.
+   */
+  protected function setBarista() {
+    $config = $this->configFactory->get('draco_coffee.settings');
+    $candidates = $this->entityTypeManager->getStorage('user')
+      ->getQuery()
+      ->condition('status', 1)
+      ->condition('roles', $config->get('role'))
```

Comment:
Is the config 'role' single or multiple? The assignment says "A select field listing user roles" and I wonder if we should simplify that to "A select field to select the role..."

Reply:
This evaluates to roles = role. For multiple roles you would use an array and IN.

Reply:
Changing the assignment as you suggest...